### ISSUE TITLE
Memory Raises a lot on deploy and does not return to normal valies

### ISSUE THREAD

**Issue by soloam** [node-red/node-red#2214]
### What are the steps to reproduce?
Don't know the exact cause, but I have a system running for weeks and the memory consumption stays stable on 2660 MB. When I start making any change (minor changes, no new logic) and start hitting the deploy the memory starts raising a lot, for example to 5G, and stays there, never goes down again. If I keep making deploys, eventually my system will crash of lack of memory.

### What happens?
Memory starts raising a lot, for example to 5G, and stays there, never goes down again. If I keep making deploys, eventually my system will crash of lack of memory.

### What do you expect to happen?
Return to normal values after the deploy.

### Please tell us about your environment:

- [ ] Node-RED version: v0.20.3
- [ ] node.js version: v8.11.3
- [ ] npm version: v6.3.0
- [ ] Platform/OS: FreeBSD / Freenas Jail
- [ ] Browser: Chrome

--------------------

Adicional information, 

- The problem started when I made the update to 0.20, on prior versions I didn't had issues 
- I have a lot of subflows in my system, and a lot of them nested.
- I have tcp connections inside subflows, I don't know if it's related, but I get a very large delay on deploys, and sometimes seems related to tcp connections.
- I have a lot of MQTT connections, and most are inside subflows,
- I run the nodered with PM2, but the problem is the same running "nodered" in forground
- The memory spikes only happen when I start making deploys, If i don't deply the system stays in low memory for weeks

Thank You

**Comment by knolleary** [comment]
Hi - thanks for raising an issue.

I note you are on 0.20.3, which isn't quite the latest. I can't think of any specific change that has since gone in for 0.20.5, but it may be worth trying if you are able to.

Can you quantify your use of subflows a bit more? Particularly interested in how you have them nested.

**Comment by dceejay** [comment]
Also do you have any 3 rd party nodes installed ? If so, which ones ?

**Comment by soloam** [comment]
@knolleary do you have any way for me to send you my complete nodered flows? I think that it's the best way to make it possible to test it out. This is the second time that I open a issue about this, and it's hard to reproduce this! I hope that with my flows you can find something useful to debug and solve a possible bug in the system.

@dceejay 

- node-red-contrib-cast (not using atm)
- node-red-contrib-gc (installed to try to debug this issue)
- node-red-contrib-gc-trigger (installed to try to debug this issue)
- node-red-contrib-home-assistant-websocket
- node-red-contrib-ifttt
- node-red-contrib-influxdb
- node-red-contrib-join-joaoapps
- node-red-contrib-lametric-notification
- node-red-contrib-merge-topic
- node-red-contrib-moment
- node-red-contrib-msg-resend
- node-red-contrib-mysensors
- node-red-contrib-npm (not using atm)
- node-red-contrib-repeat
- node-red-contrib-samsungTV
- node-red-contrib-telegrambot
- node-red-contrib-time-range-switch
- node-red-contrib-viera
- node-red-dashboard
- node-red-node-emoncms
- node-red-node-openweathermap
- node-red-node-sentiment

**Comment by dceejay** [comment]
hmm - yeah a good list... any of which may not be closing down properly before a redeploy...
I would certainly try removing any that you aren't using and then retrying. The fewer you have in use the more likely we are to be able to help... We only really support the core (node-red-node) ones and don't really have the bandwidth to help debug every possible extra node.

**Comment by soloam** [comment]
I tried starting to remove them and the problem still persists, that was my
first approach

**Comment by knolleary** [comment]
If the only thing that has changed was you upgraded to 0.20.x, then I wouldn't rush to point the finger at a 3rd party node quite yet.

The fact there are lots of subflows and nested subflows is interesting as there were significant changes to the internal handling of subflows in 0.20 and we need to make sure that hasn't regressed something here. Our unit tests dont show any issues, but they don't cover every possible scenario.

Getting a better sense of the pattern of subflows you have will have is see if you are hitting an edge case our tests dont.

I'd rather avoid having to debug you full flow if you can provide a more succinct summary of what you have. Failing that, if you join the slack channel (nodered.org/slack) you can send me your flow file there.

**Comment by soloam** [comment]
If I rollback to 0.19 the problem is solved! I would prefere not to,
because I really like the new subflow functionality and I would lose the
test case.

My main functionality of nodered is to define my devices and then I have a
subflow to the different type of brands one for in, and one for out! I
recive the actions by mqtt and I do the response to the devices, some are
mqtt, others are tcp or http!

Basically I must have around 115 subflows (🙄) and most of them call other
in stack, I would say that the top should be 5 stack, probably more!

I tried disable some flows and the problem still persists! So I think that
the problem is related to the subflows! Please note that if I don't deploy
the memory does not raise, so I ruled out memory leaks in my code!

Thank you so much for your help

**Comment by knolleary** [comment]
115 subflows .... 🙈

So if you can send me your flow file via slack, that would be very helpful

**Comment by knolleary** [comment]
Thank you for sharing your flows with me.

As I don't have all of the 3rd party nodes that you have - nor do I want to test your flow logic - I replaced all of your flow nodes with Change nodes and all of your config nodes with `tls-config` nodes. This meant the flows would run, but no messages would flow.

I then ran a few tests to see what was going on, using OSX Activity Monitor to get rough memory usage figures.

With 0.19.5, node-red sits around 170Mb when started. I then did 5  'move node, deploy change' cycles - with no real pause between them. The memory usage crept up to a peak of 350Mb and then within a minute or two had dropped back to 170Mb.

With 0.20.5, node-red starts around 437Mb but settled back to 380Mb. I then did the 5 'move node, deploy change' cycles. Memory usage peaked at 1.03Gb. The deploys also felt like they took longer.
After a minute, memory dropped back to 350Mb.


On that basis, I don't believe there is an obvious memory leak in the core. I do see the memory usage drop back to 'normal' after a series of deploys.

But there clearly has been a huge increase in the memory Node-RED needs to run. That is unexpected and needs to be investigated.

**Comment by soloam** [comment]
Hello @knolleary thank you for your time

You behavior looks almost like mine, the main diference is the drop back on memory. On my case the systems crashes, probably because the raising is a bigger.

So can you will look in to this and try to find the problem? Can I do anything to help on this?

Thank you

**Comment by soloam** [comment]
Hello @knolleary I think I discovered the root of the problems of my flows! As I suspected, it's related with the subflows. The main problem is that I have a subflow called "Devices Gateway Out"... This subflow is the main gateway from all my devices from nodered, and I use it to talk with the devices... Inside this Gateway I have several subflows that talk with the specific protocols, for example Sonoff, Air Condicioners, etc etc... This subflow is inside a subflow called "Save Device" That saves the device info and calls the "Gateway" to communicate with the device.

The problem is that this "Save Device" subflow is in a lot of places, and if I remove the "Gateway" from inside the node, my memory goes from 4000Mb To 700Mb... Should this happen? Should't be only one instance of "Gateway" inside all the "Save Device"? This was not a problem before 0.20.

Thank You

**Comment by soloam** [comment]
I made a very small change, Inside my "Save Device" I removed the "Gateway" and added JSON stringify() from the complete message to inside the payload, and submitted to a topic. Inside the main flow I recive the message, do a JSON.parse() to a new object and return the message, it's the exact same message, now I send it to the "Gateway" and my flows work as before. The main diference is that before my nodered was conuming 4GB of ram, now only 569Mb.

Problem spotted :)... Now I think that this should not be a problem... From what I see my "Gateway" subnode should not be spammed across all "Save Devices" nodes... should he?

Thank You

**Comment by knolleary** [comment]
> Should't be only one instance of "Gateway" inside all the "Save Device"? This was not a problem before 0.20.

No. Each instance of a subflow you add is a separate copy of the subflow. This is how subflows have always worked.

If you have Subflow A that contains 2 copies of Subflow B, and Subflow B contains two copies of Subflow C, then when you add one instance of the A to your workspace you will end up with 1xA, 2xB and 4xC.

The changes we made in 0.20 have certainly increased the memory usage of each subflow instance - something we need to investigate. I think your particular arrangement of many nested subflows is resulting in many more instances of subflows then you realise. Given the increased memory usage, that is compounding the problem.
